### PR TITLE
[feat]: [PL-28387]: In AccessControlClient, added new overload method for performing RBAC check.

### PR DESCRIPTION
**Pull request by Jimit09Gandhi**
## Describe your changes
In AccessControlClient, added new overload method for performing RBAC check.
This method takes in List of  permissionCheckDTOs and makes call to ACL in batches of 1000 and abstracts the need to the same for every harness resource entity and throwing access denied exception in case Principal does not have the access to even one. This will be particularly useful for List APIs performing RBAC check.

## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>

You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
    - CodeFormat: `trigger codeformat`
    - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- Trigger all Checks: `trigger smartchecks`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/40771)

### CODE REVIEW

File: clients/access-control/sdk/src/main/java/io/harness/accesscontrol/clients/AbstractAccessControlClient.java [hunk 1]
```diff
@@ -108,6 +115,25 @@ public void checkForAccessOrThrow(
     checkForAccessOrThrowInternal(principal, resourceScope, resource, permission, exceptionMessage);
   }
 
+  @Override
+  public AccessCheckResponseDTO checkForAccessOrThrow(List<PermissionCheckDTO> permissionCheckDTOList) {
```
Comment (meenakshi-harness) [inline clients/access-control/sdk/src/main/java/io/harness/accesscontrol/clients/AbstractAccessControlClient.java:119]: UTs for this?

File: clients/access-control/sdk/src/main/java/io/harness/accesscontrol/clients/AbstractAccessControlClient.java [hunk 2]
```diff
@@ -108,6 +115,25 @@ public void checkForAccessOrThrow(
...
+            .collect(Collectors.toList());
+    List<AccessControlDTO> accessControlList = new ArrayList<>();
+    accessCheckResponseDTOs.forEach(res -> accessControlList.addAll(res.getAccessControlList()));
+    if (isNotEmpty(accessControlList) && accessControlList.stream().noneMatch(AccessControlDTO::isPermitted)) {
+      String message = String.format("Missing permission %s on %s", accessControlList.get(0).getPermission(),
+          accessControlList.get(0).getResourceType().toLowerCase());
+      throw new NGAccessDeniedException(message, USER, emptyList());
+    }
+    return AccessCheckResponseDTO.builder()
+        .principal(accessCheckResponseDTOs.get(0).getPrincipal())
```
Comment (meenakshi-harness) [inline clients/access-control/sdk/src/main/java/io/harness/accesscontrol/clients/AbstractAccessControlClient.java:132]: Should we check accessCheckResponseDTOs is not empty?

**Review by meenakshi-harness — CHANGES_REQUESTED** [review]

**Review by Jimit09Gandhi — COMMENTED** [review]
> Reviewed 1 of 2 files at r1, 3 of 3 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/40771)* status: 4 file(s) to review. 1 change requests. 2. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`clients/access-control/sdk/src/main/java/io/harness/accesscontrol/clients/AbstractAccessControlClient.java` line 119 at r1](https://reviewable.io/reviews/harness/harness-core/40771#-NHzX-JbJqHd_d6Zgmw_-r1-119:-NIDprRH-ShM7PWQSze8:bqvw7wv) ([raw file](https://github.com/harness/harness-core/blob/f9b9fe62e428ba6be986a26fb2e9c7ed5f509ec5/clients/access-control/sdk/src/main/java/io/harness/accesscontrol/clients/AbstractAccessControlClient.java#L119)):*
<details><summary><i>Previously, meenakshi-harness (Meenakshi Raikwar) wrote…</i></summary><blockquote>

UTs for this?
</blockquote></details>

Added UTs
___
*[`clients/access-control/sdk/src/main/java/io/harness/accesscontrol/clients/AbstractAccessControlClient.java` line 132 at r1](https://reviewable.io/reviews/harness/harness-core/40771#-NHzX-JbJqHd_d6Zgmw_-r1-132:-NIDpubk7dYCHcBFi_r9:b-pre4ts) ([raw file](https://github.com/harness/harness-core/blob/f9b9fe62e428ba6be986a26fb2e9c7ed5f509ec5/clients/access-control/sdk/src/main/java/io/harness/accesscontrol/clients/AbstractAccessControlClient.java#L132)):*
<details><summary><i>Previously, meenakshi-harness (Meenakshi Raikwar) wrote…</i></summary><blockquote>

Should we check accessCheckResponseDTOs is not empty?
</blockquote></details>

Added check on isEmpty check on permissionCheckDTOList. It is expected that accessCheckResponseDTOs will no of entities corresponding to permissionCheckDTOList.

File: clients/access-control/sdk/BUILD.bazel [hunk 3]
```diff
@@ -27,10 +27,15 @@ java_library(
         "//980-commons:module",
         "//990-commons-test:module",
         "//999-annotations:module",
+        "//access-control/contracts:module",
+        "//clients/access-control/sdk:module",
```
Comment (reetika04) [inline clients/access-control/sdk/BUILD.bazel:31]: Isn’t this already "//clients/access-control/sdk:module"?

**Review by reetika04 — COMMENTED** [review]
> Reviewed 1 of 2 files at r1, 3 of 3 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/40771)* status: 4 file(s) to review. 1 change requests. 2. unresolved discussions. 0 of 2 approvals obtained.

**Comment by Jimit09Gandhi** [comment]
trigger checkstyle

**Review by meenakshi-harness — APPROVED** [review]
> Reviewed 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/40771)* status: 1 file(s) to review.

**Review by meenakshi-harness — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/40771)* status: 1 file(s) to review.

**Review by reetika04 — COMMENTED** [review]
> Reviewed 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/40771)* status: :shipit: complete! ✓

**Review by reetika04 — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/40771)* status: :shipit: complete! ✓

**Review by ankit-harness — APPROVED** [review]
:lgtm:

> Reviewed 1 of 2 files at r1, 2 of 3 files at r2, 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/40771#-:-NJ3uWRp994XytzJZYdo:bnfp4nl)* status: Some of the checks are too old.

**Pull request closed without merging**